securedrop/tests/common.py

```diff
@@ -1,120 +1,272 @@
+    def init_gpg():
+        # Initialize the GPG keyring
+        gpg = gnupg.GPG(homedir=config.GPG_KEY_DIR)
+        # Import the journalist key for testing (faster to import a pre-generated
+        # key than to gen a new one every time)
+        for keyfile in ("test_journalist_key.pub", "test_journalist_key.sec"):
+            gpg.import_keys(open(keyfile).read())
+        return gpg
+
+    @classmethod
+    def setup(cls):
+        """Set up the file system, GPG, and database."""
+        SetUp.create_directories()
+        SetUp.init_gpg()
+        init_db()
+        # Do tests that should always run on app startup
+        crypto_util.do_runtime_tests()
+        # Start the Python-RQ worker if it's not already running
+        if not os.path.exists(cls.test_worker_pidfile):
+            subprocess.Popen(["rqworker",
+                              "-P", config.SECUREDROP_ROOT,
+                              "--pid", cls.test_worker_pidfile])
+
 
-    return filenames
+class TearDown:
+    """Test fixtures for tearing down a generic test environment.
+    """
+    @staticmethod
+    def teardown():
+        shutil.rmtree(config.SECUREDROP_DATA_ROOT)
 
+    # TODO: now that SD has a logout button, we can deprecate use of
+    # this function.
+    @staticmethod
+    def logout(test_client):
+        with test_client.session_transaction() as sess:
+            sess.clear()
 
-def new_codename(client, session):
-    """Helper function to go through the "generate codename" flow"""
-    with client as c:
-        rv = c.get('/generate')
-        codename = session['codename']
-        rv = c.post('/create')
-    return codename
 
+class Async:
+    """Test fixtures for use with asynchronous processes.
+    """
+    redis_success_return_value = 'success'
 
-def shared_setup():
-    """Set up the file system, GPG, and database"""
-    create_directories()
-    init_gpg()
-    init_db()
+    @classmethod
+    def wait_for_redis_worker(cls, job, timeout=5):
+        """Raise an error if the Redis job doesn't complete successfully
+        before a timeout.
 
-    # Do tests that should always run on app startup
-    crypto_util.do_runtime_tests()
+        :param rq.job.Job job: A Redis job to wait for.
 
-    # Start the Python-RQ worker if it's not already running
-    if not os.path.exists(TEST_WORKER_PIDFILE):
-        subprocess.Popen(["rqworker", "-P", config.SECUREDROP_ROOT,
-                                      "--pid", TEST_WORKER_PIDFILE])
+        :param int timeout: Seconds to wait for the job to finish.
 
+        :raises: An :exc:`AssertionError`.
+        """
+        start_time = time.time()
+        while time.time() - start_time < timeout:
+            if job.result == cls.redis_success_return_value:
+                return
+            elif job.result not in (None, cls.redis_success_return_value):
+                assert False, 'Redis worker failed!'
+            time.sleep(0.1)
+        assert False, 'Redis worker timed out!'
```

Comment:
That makes sense when caller functions need to perform exception handling based on what goes wrong. In this case the test runner will just fail the unit test and the message to the developer will be sufficiently clear.